test/workflow_test.go

```diff
@@ -1356,6 +1356,55 @@ func (w *Workflows) SleepForDuration(ctx workflow.Context, d time.Duration) erro
 	return workflow.Sleep(ctx, d)
 }
 
+func (w *Workflows) ActivityPostCancellation(ctx workflow.Context) error {
+	waitingForCancel := false
+	err := workflow.SetQueryHandler(ctx, "waiting-for-cancel", func() (bool, error) {
+		return waitingForCancel, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
+		StartToCloseTimeout: 5 * time.Minute,
+		HeartbeatTimeout:    5 * time.Second,
+		WaitForCancellation: true,
+	})
+	var a *Activities
+
+	// Run activity to cancel
+	fut := workflow.ExecuteActivity(ctx, a.HeartbeatUntilCanceled, 1*time.Second)
+	waitingForCancel = true
+	if err := fut.Get(ctx, nil); err != nil {
+		return fmt.Errorf("activity did not gracefully cancel: %w", err)
+	}
+
+	// Run another activity after cancel
+	ctx, _ = workflow.NewDisconnectedContext(ctx)
+	return workflow.ExecuteActivity(ctx, a.Sleep, 1*time.Millisecond).Get(ctx, nil)
+}
+
+func (w *Workflows) TimerPostCancellation(ctx workflow.Context) error {
```

Comment:
I think another useful test to add would be both activity and timer with this pattern.

Possibly one with more than one of each type as well

Reply:
I have added all the permutations. It does add 5 seconds onto integration tests unfortunately.